cdc/sink/cdclog/file.go

```diff
@@ -0,0 +1,338 @@
+// limitations under the License.
+
+package cdclog
+
+import (
+	"context"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/uber-go/atomic"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	parsemodel "github.com/pingcap/parser/model"
+
+	"github.com/pingcap/ticdc/cdc/model"
+	"github.com/pingcap/ticdc/cdc/sink/codec"
+)
+
+const (
+	defaultDirMode  = 0755
+	defaultFileMode = 0644
+
+	maxRowFileSize = 10 << 20 // TODO update
+)
+
+type logPath struct {
+	root string
+	ddl  string
+	meta string
+}
+
+type tableStream struct {
+	dataCh  chan *model.RowChangedEvent
+	rowFile *os.File
+
+	tableID    int64
+	sendEvents *atomic.Int64
+	sendSize   *atomic.Int64
+}
+
+func newTableStream(tableID int64) *tableStream {
+	return &tableStream{
+		tableID: tableID,
+		dataCh:  make(chan *model.RowChangedEvent, defaultBufferChanSize),
+
+		sendEvents: atomic.NewInt64(0),
+		sendSize:   atomic.NewInt64(0),
+	}
+}
+
+type fileSink struct {
+	logMeta *logMeta
+	logPath *logPath
+
+	ddlFile *os.File
+	encoder func() codec.EventBatchEncoder
+
+	hashMap      sync.Map
+	tableStreams []*tableStream
+}
+
+func (f *fileSink) flushTableStreams(ctx context.Context) error {
+	// TODO use a fixed worker pool
+	eg, _ := errgroup.WithContext(ctx)
+	for _, ts := range f.tableStreams {
+		tsReplica := ts
+		eg.Go(func() error {
+			var fileName string
+			flushedEvents := tsReplica.sendEvents.Load()
+			flushedSize := tsReplica.sendSize.Load()
+			encoder := f.encoder()
+			for event := int64(0); event < flushedEvents; event++ {
+				row := <-tsReplica.dataCh
+				if event == flushedEvents-1 {
+					// the last event
+					fileName = makeTableFileName(row.CommitTs)
+				}
+				_, err := encoder.AppendRowChangedEvent(row)
+				if err != nil {
+					return err
+				}
+			}
+			rowDatas := encoder.MixedBuild()
+
+			log.Debug("[flushTableStreams] build cdc log data",
+				zap.Int64("table id", tsReplica.tableID),
+				zap.Int64("flushed size", flushedSize),
+				zap.Int64("flushed event", flushedEvents),
+				zap.Int("encode size", len(rowDatas)),
+				zap.String("file name", fileName),
+			)
+
+			tableDir := filepath.Join(f.logPath.root, makeTableDirectoryName(tsReplica.tableID))
+
+			if tsReplica.rowFile == nil {
+				// create new file to append data
+				err := os.MkdirAll(tableDir, defaultDirMode)
+				if err != nil {
+					return err
+				}
+				file, err := os.OpenFile(filepath.Join(tableDir, fileName), os.O_CREATE|os.O_WRONLY|os.O_APPEND, defaultFileMode)
+				if err != nil {
+					return err
+				}
+				tsReplica.rowFile = file
+			}
+
+			stat, err := tsReplica.rowFile.Stat()
+			if err != nil {
+				return err
+			}
+
+			if stat.Size() > maxRowFileSize {
+				// rotate file
+				err := tsReplica.rowFile.Close()
+				if err != nil {
+					return err
+				}
+				file, err := os.OpenFile(filepath.Join(tableDir, fileName), os.O_CREATE|os.O_WRONLY|os.O_APPEND, defaultFileMode)
```

Comment:
No, when every batch flush, it will create new file name [here](https://github.com/pingcap/ticdc/pull/826/files/9f72844e79dcb885735803fd8a03e252333acaca#diff-2662504f1f959169dbcc1c8723c19361R93), but not every new file name will be used.

Reply:
eh... maybe we should use the first event' commit ts as the filename